javatests/arcs/core/analysis/RecipeGraphFixpointIteratorTest.kt

```diff
@@ -0,0 +1,267 @@
+) : RecipeGraphFixpointIterator<AbstractSet<String>>() {
+    override fun getInitialValues(graph: RecipeGraph) = graph.nodes.filter {
+        startNames.contains(it.getName())
+    }.associateWith {
+        AbstractSet<String>(setOf())
+    }
+
+    override fun transform(handle: Recipe.Handle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "h:${handle.name}") } ?: input
+
+    override fun transform(particle: Recipe.Particle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "p:${particle.spec.name}") } ?: input
+
+    override fun transform(
+        handle: Recipe.Handle,
+        particle: Recipe.Particle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "h:${handle.name} -> p:${particle.spec.name}")
+    } ?: input
+
+    override fun transform(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "p:${particle.spec.name} -> h:${handle.name}")
+    } ?: input
+}
+
+@RunWith(JUnit4::class)
+class RecipeGraphFixpointIteratorTest {
+    private val thing = Recipe.Handle("thing", Recipe.Handle.Fate.CREATE, TypeVariable("thing"))
+    private val name = Recipe.Handle("name", Recipe.Handle.Fate.CREATE, TypeVariable("name"))
+    private val readCnxn = HandleConnectionSpec("r", HandleMode.Read, TypeVariable("r"))
+    private val writeCnxn = HandleConnectionSpec("w", HandleMode.Write, TypeVariable("w"))
+    private val readerSpec = ParticleSpec(
+        "Reader",
+        listOf(readCnxn).associateBy { it.name },
+        "ReaderLocation"
+    )
+    private val writerSpec = ParticleSpec(
+        "Writer",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val anotherWriterSpec = ParticleSpec(
+        "AnotherWriter",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val readerParticle = Recipe.Particle(
+        readerSpec,
+        listOf(Recipe.Particle.HandleConnection(readCnxn, thing))
+    )
+    private val writerParticle = Recipe.Particle(
+        writerSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+    private val anotherWriterParticle = Recipe.Particle(
+        anotherWriterSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+
+    @Test
+    fun straightLineFlow() {
+        // [Writer] -> (thing) -> [Reader]
+        val recipe = Recipe(
+            "StraightLine",
+            listOf(thing).associateBy { it.name },
+            listOf(readerParticle, writerParticle),
+            "arcId"
+        )
+        val graph = RecipeGraph(recipe)
+        val analyzer = TestAnalyzer(setOf("p:${writerParticle.spec.name}"))
+        analyzer.computeFixpoint(graph)
```

Comment:
Nit: can you add a blank line above and below this line, to separate out the Arrange, Act, Assert blocks more clearly? Here and throughout file.

Reply:
Done.